Add SDL2 support

deveee opened the pull request:
If you build it, make sure that _IRR_COMPILE_WITH_SDL2_DEVICE_ is enabled either as a compiler flag or in IrrCompileConfig.h

As far as I tested, Irrlicht examples work fine with SDL2 on Linux, Android, Windows and MacOS. Supported drivers are OpenGL, GLES2 and GLES1.

Compared to other drivers, one missing feature is custom mouse cursor support.

Currently Visual Studio, Xcode, Codeblocks, MinGW projects are not modified, so you have to manually add SDL library and include directories to the project.

MoNTE48 commented:
I think [Add example for SDL2 on Android](https://github.com/MoNTE48/Irrlicht/pull/1/commits/6209f0a605bec725664229e233f98c4dc41872a6) commit can be dropped.

deveee committed:
Add SDL2 support

deveee committed:
Some fixes for iOS

deveee committed:
Make sure that framebuffer is initialized to default value on iOS

deveee committed:
Add high dpi support for iOS

deveee committed:
Fixed mouse warp with high dpi

deveee committed:
Send shift and control state in mouse events

deveee committed:
Use command key as control modifier on mac

deveee committed:
Mark key events without char as handled.

SDL2 has saparate textinput events and Char in key event is always 0.

deveee committed:
Stop text input events on delete only if it was started by this edit box.

deveee committed:
Make it working with non-latin keyboards

deveee committed:
Add SDL2 window to exposed video data

MoNTE48 committed:
Spaces

deveee committed:
Read native scale before actual window is created.

Also handle the case when width or height is equal to 0

deveee committed:
Another way to read high dpi scale without hidden window creation

deveee committed:
Try to use smooth scroll

MoNTE48 committed:
Direct3D O_o

MoNTE48 committed:
fix code style a little bit

deveee committed:
Try to enable AppleMomentumScrollSupported.

Also send both horizontal and vertical events.

MoNTE48 committed:
Hmm... X scrolling fixed?

MoNTE48 committed:
to be sure.

MoNTE48 committed:
Little code format

MoNTE48 committed:
Remove unsupported by SDL2 Software and Burning Video, DeviceConsole and DeviceiOS/OSX

MoNTE48 committed:
Remove double added `CIrrDeviceSDL2.cpp`

MoNTE48 commented:
@deveee After this commit https://github.com/MoNTE48/Irrlicht/pull/1/commits/c9dfd658c89ce00d41b2da03bc78db019325c7a5 I get the next error when starting MultiCraft:
error build: Undefined symbol: irr::CIrrDeviceMacOSX::CIrrDeviceMacOSX(irr::SIrrlichtCreationParameters const&)

deveee commented:
Because both SDL2 and OSX devices were built (it's possible to select irrlicht device using DriverType in creation params). This should disable OSX device if compiled with SDL2 (I can test later).
```
diff --git a/include/IrrCompileConfig.h b/include/IrrCompileConfig.h
index 487f35c1..fd484bd0 100644
--- a/include/IrrCompileConfig.h
+++ b/include/IrrCompileConfig.h
@@ -119,7 +119,9 @@
 #define NO_IRR_COMPILE_WITH_OPENGL_
 #else
 #define _IRR_OSX_PLATFORM_
+#if !defined(_IRR_COMPILE_WITH_SDL2_DEVICE_)
 #define _IRR_COMPILE_WITH_OSX_DEVICE_
+#endif
 #define NO_IRR_COMPILE_WITH_OGLES1_
 #define NO_IRR_COMPILE_WITH_OGLES2_
 #define NO_IRR_COMPILE_WITH_WEBGL1_
```

deveee committed:
Don't build OSX device if SDL2 is enabled

deveee committed:
Define _IRR_COMPILE_WITH_SDL2_DEVICE_ in xcode project

deveee committed:
Build for iOS without SDL2 for now and remove unused file from project

deveee committed:
One more tweak...

deveee committed:
Try to create GLES 3.0 context first and fallback to GLES 2.0 if not success.

iOS doesn't give highest possible context when GLES 2.0 is requested and
thus npot textures don't work.

deveee committed:
Disable start/stop text input on editbox focus change by default

deveee committed:
Rename to SDL device

MoNTE48 commented on source/Irrlicht/Irrlicht.cpp:
```diff
@@ -73,6 +73,11 @@ namespace irr
+		if (params.DeviceType == EIDT_SDL || (!dev && params.DeviceType == EIDT_BEST))
+			dev = new CIrrDeviceSDL(params);
+#endif
+
```
Can you not move it up?

deveee committed:
Keep original device order in Irrlicht.cpp

MoNTE48 committed:
Update CIrrDeviceSDL.cpp
